### PR TITLE
Respond to parallel replicas protocol requests with priority on initiator

### PR DESCRIPTION
**Pull request by devcrafter**
This is necessary for local plan execution with PR. W/o this change, responses to followers are delayed by local query execution (see [comment](https://github.com/ClickHouse/ClickHouse/pull/66618/files#diff-61b189d4068bc450f6bd7784930921f20280da12a215faf9f0b82e0b42afa739R234-R237))

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit b6a790124cd670749b4c504f58a4854307bf7d83 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66618/b6a790124cd670749b4c504f58a4854307bf7d83/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66586/e8627601b52e28a4d4203cd365cd5e4693e2af57/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/QueryPipeline/RemoteQueryExecutor.cpp [hunk 1]
```diff
@@ -914,4 +924,41 @@ bool RemoteQueryExecutor::needToSkipUnavailableShard() const
     return context->getSettingsRef().skip_unavailable_shards && (0 == connections->size());
 }
 
+bool  RemoteQueryExecutor::processParallelReplicaPacketIfAny()
```
Comment (pufit) [inline src/QueryPipeline/RemoteQueryExecutor.cpp:927]:
```suggestion
bool RemoteQueryExecutor::processParallelReplicaPacketIfAny()
```

File: src/QueryPipeline/RemoteQueryExecutor.cpp [hunk 2]
```diff
@@ -914,4 +924,41 @@ bool RemoteQueryExecutor::needToSkipUnavailableShard() const
...
+#if defined(OS_LINUX)
+    if (!read_context || (resent_query && recreate_read_context))
+    {
+        std::lock_guard lock(was_cancelled_mutex);
+        if (was_cancelled)
+            return false;
+
+        read_context = std::make_unique<ReadContext>(*this);
+        recreate_read_context = false;
+    }
```
Comment (pufit) [inline src/QueryPipeline/RemoteQueryExecutor.cpp:938]: why do we need to release the mutext just to retake it two lines below?

File: src/Processors/Sources/RemoteSource.cpp [hunk 3]
```diff
@@ -98,9 +98,26 @@ void RemoteSource::work()
...
+    if (preprocessed_packet)
+    {
+        preprocessed_packet = false;
+        return;
+    }
+
     ISource::work();
 }
 
+void RemoteSource::onAsyncJobReady()
```
Comment (pufit) [inline src/Processors/Sources/RemoteSource.cpp:111]: this function is unused (?)

File: src/Processors/IProcessor.h [hunk 4]
```diff
@@ -221,6 +221,23 @@ class IProcessor
...
+     * See also comment to schedule() method
+     *
+     * It allows doing some preprocessing immediately after asynchronous job is done.
+     * The implementation should return control quickly, to avoid blocking another asynchronous completed jobs
+     * created by the same pipeline.
+     *
+     * Example, scheduling tasks for remote workers (file descriptor in this case is a socket)
+     * When the remote worker asks for the next task, doing it in onAsyncJobReady() we can provide it immediately.
+     * Otherwise, the returning of the next task for the remote worker can be delayed by current work done in the pipeline
+     * i.e. processor->work(), which will create unnecessary latency in query processing by remote workers   Not Committed Yet
```
Comment (serxa) [inline src/Processors/IProcessor.h:237]:
```suggestion
     * i.e. processor->work(), which will create unnecessary latency in query processing by remote workers
```

**Review by serxa — APPROVED** [review]

**Comment by devcrafter** [comment]
`03001_restore_from_old_backup_with_matview_inner_table_metadata` failure in https://s3.amazonaws.com/clickhouse-test-reports/66618/3ba8c3166b41a3a549d41cfa2d65b81ea7be6db7/stateless_tests__debug__s3_storage__[2_2].html is not related

The test failure is not frequent, there are a couple of other [failures](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDM2MCBIT1VSCi0tICAgIEFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIHRlc3Rfc3RhdHVzICE9ICdTS0lQUEVEJwogICAgQU5EICh0ZXN0X3N0YXR1cyBMSUtFICdGJScgT1IgdGVzdF9zdGF0dXMgTElLRSAnRSUnKSAKICAgIEFORCBjaGVja19zdGF0dXMgIT0gJ3N1Y2Nlc3MnCiAgICBBTkQgY2hlY2tfbmFtZSBOT1QgTElLRSAnU1FMYW5jZXIlJwogICAgQU5EIHRlc3RfbmFtZSBpbGlrZSAnJTAzMDAxX3Jlc3RvcmVfZnJvbV9vbGRfYmFja3VwX3dpdGhfbWF0dmlld19pbm5lcl90YWJsZV9tZXRhZGF0YSUnCk9SREVSIEJZIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgY2hlY2tfc3RhcnRfdGltZQ==) for this test other PRs

cc @vitlibar

**Pull request merged**